src/error-handling/error-contexts.md

```diff
@@ -5,25 +5,18 @@ contextual information to your errors and allows you to have fewer
 custom error types:
 
 ```rust,editable,compile_fail
-use std::{fs, io};
+use std::{fs::{self, File}, io};
 use std::io::Read;
-use thiserror::Error;
-use anyhow::{Context, Result};
-
-#[derive(Error, Debug)]
-enum ReadUsernameError {
-    #[error("Found no username in {0}")]
-    EmptyUsername(String),
-}
+use anyhow::{Context, Result, bail};
 
 fn read_username(path: &str) -> Result<String> {
     let mut username = String::with_capacity(100);
-    fs::File::open(path)
+    File::open(path)
         .context(format!("Failed to open {path}"))?
         .read_to_string(&mut username)
         .context("Failed to read")?;
     if username.is_empty() {
-        return Err(ReadUsernameError::EmptyUsername(String::from(path)))?;
```

Comment:
With this change, it's no longer possible to downcast to the original error, right?

I'm not sure why this is a good idea then :slightly_smiling_face:

Reply:
True, but if you want to distinguish different types of errors programmatically you probably shouldn't be using `anyhow`. Usually people use `anyhow` or similar dynamic error types because they just want to print an error message or backtrace rather than checking details of the error in code. I think using `anyhow!` or `bail!` is more idiomatic than defining a custom error type only to immediately wrap it in an `anyhow::Error`.